### PR TITLE
XT-3408: NL FG-NL-11 validation

### PR DESCRIPTION
**Pull request by allenbross-wf**
#### Reason for change

NL FG-NL-04 validation

#### Description of change

Added NL FG-NL-04 validation.

#### Steps to Test

CI

**review**:
@Arelle/arelle

### CODE REVIEW

**Comment by aviary2-wf** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: arelle/plugin/validate/NL/rules/fg_nl.py [hunk 1]
```diff
@@ -92,3 +97,38 @@ def rule_fg_nl_04(
...
+) -> Iterable[Validation] | None:
+    """
+    FG-NL-11: XBRL instance file Should only contain standard units based on the XBRL units Registry.
+    """
+    modelXbrl = val.modelXbrl
+    customUnits = []
+    for unitId, unit in modelXbrl.units.items():
+        for measures in unit.measures:
+            for measure in measures:
+                if measure.namespaceURI not in VALID_UTR_NAMESPACES:
```
Comment (brettkail-wk) [inline arelle/plugin/validate/NL/rules/fg_nl.py:124]: This seems to allow `unitRef="iso4217:brett"`, etc.

If the conclusion is to use a simple namespace-based test, it seems like we should add FIXME-style comment saying that loading from some specific version of the UTR would be more precise.

File: arelle/plugin/validate/NL/rules/fg_nl.py [hunk 2]
```diff
@@ -92,3 +97,38 @@ def rule_fg_nl_04(
...
+                    customUnits.append(unit)
+    if customUnits:
+        for unit in customUnits:
+            yield Validation.warning(
+                codes='NL.FG-NL-11',
+                msg=_('"Default units SHOULD be used in measure elements, a.k.a. should not be using a custom unit. '
```
Comment (brettkail-wk) [inline arelle/plugin/validate/NL/rules/fg_nl.py:130]: The leading `"` looks wrong.

"Default unit" is an unfamiliar term, and "also known as" seems like odd phrasing.  Perhaps it should just say what the manual says: "Standard units SHOULD be used"

File: arelle/plugin/validate/NL/rules/fg_nl.py [hunk 3]
```diff
@@ -35,6 +35,11 @@
     'fact': 6,
 }
 
+VALID_UTR_NAMESPACES = [
+    'http://www.xbrl.org/2003/instance',
+    'http://www.xbrl.org/2009/utr',
+    'http://www.xbrl.org/2003/iso4217'
+]
```
Comment (aaroncameron-wk) [inline arelle/plugin/validate/NL/rules/fg_nl.py:42]: Going to be a flake8 issue here, need to retain two lines of whitespace above function definition.

File: tests/resources/conformance_suites/nl_nt16/fg_nl/11-custom-unit.xbrl [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        </context>
+    <xbrli:unit id="mky">
+        <xbrli:measure>mock:monkey</xbrli:measure>
+    </xbrli:unit>
+    <xbrli:unit id="usd">
+        <xbrli:measure>iso4217:USD</xbrli:measure>
+    </xbrli:unit>
+    <mock:Monetary contextRef="ctx-1" unitRef="mky" decimals="INF">100</mock:Monetary>
+   <mock:Monetary contextRef="ctx-2" unitRef="usd" decimals="INF">100</mock:Monetary>
+</xbrl>
```
Comment (aaroncameron-wk) [inline tests/resources/conformance_suites/nl_nt16/fg_nl/11-custom-unit.xbrl:39]: Nit: EOF newline

File: tests/resources/conformance_suites/nl_nt16/fg_nl/11-testcase.xml [hunk 5]
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (aaroncameron-wk) [inline tests/resources/conformance_suites/nl_nt16/fg_nl/11-testcase.xml:1]: Doesn't look like these testcases were added to the corresponding `index.xml` files.

File: arelle/plugin/validate/NL/rules/fg_nl.py [hunk 6]
```diff
@@ -35,6 +35,11 @@
     'fact': 6,
 }
 
+VALID_UTR_NAMESPACES = [
+    'http://www.xbrl.org/2003/instance',
+    'http://www.xbrl.org/2009/utr',
+    'http://www.xbrl.org/2003/iso4217'
```
Comment (aaroncameron-wk) [inline arelle/plugin/validate/NL/rules/fg_nl.py:41]: We've defined these individual URL values as constants in `XbrlConst`.

File: arelle/plugin/validate/NL/rules/fg_nl.py [hunk 7]
```diff
@@ -10,8 +10,9 @@
 from arelle.ModelInstanceObject import ModelFact
 from arelle.ModelObject import ModelObject
 from arelle.ModelValue import qname, qnameEltPfxName
+from arelle.ValidateUtr import ValidateUtr
 from arelle.ValidateXbrl import ValidateXbrl
-from arelle.XbrlConst import qnLinkSchemaRef, qnLinkLinkbaseRef, qnLinkRoleRef, qnLinkArcroleRef, qnXbrliContext, qnXbrliUnit, qnEnumerationItemTypes
+from arelle.XbrlConst import iso4217, qnLinkSchemaRef, qnLinkLinkbaseRef, qnLinkRoleRef, qnLinkArcroleRef, qnXbrliContext, qnXbrliUnit, qnEnumerationItemTypes, utr, xbrli
```
Comment (aaroncameron-wk) [inline arelle/plugin/validate/NL/rules/fg_nl.py:15]: No longer using these imports

File: arelle/plugin/validate/NL/rules/fg_nl.py [hunk 8]
```diff
@@ -115,6 +116,7 @@ def rule_fg_nl_05(
         **kwargs: Any,
 ) -> Iterable[Validation] | None:
     """
+<<<<<<< Updated upstream
```
Comment (aaroncameron-wk) [inline arelle/plugin/validate/NL/rules/fg_nl.py:119]: Leftover from merge

File: arelle/plugin/validate/NL/rules/fg_nl.py [hunk 9]
```diff
@@ -193,3 +195,44 @@ def rule_fg_nl_09(
...
+    FG-NL-11: XBRL instance file Should only contain standard units based on the XBRL units Registry.
+    """
+    modelXbrl = val.modelXbrl
+    customUnits = []
+    utrItemTypeEntries = ValidateUtr(modelXbrl).utrItemTypeEntries
+    validUtrUnits = set()
+    for entries in utrItemTypeEntries.values():
+        for entry in entries.values():
+            if entry.nsUnit and entry.unitId:
+                #unitQname = entry.nsUnit.rsplit("/")[-1] + ':' + entry.unitId
```
Comment (aaroncameron-wk) [inline arelle/plugin/validate/NL/rules/fg_nl.py:224]: Commented code

File: arelle/plugin/validate/NL/rules/fg_nl.py [hunk 10]
```diff
@@ -193,3 +195,44 @@ def rule_fg_nl_09(
...
+        for measures in unit.measures:
+            for measure in measures:
+                if (measure.namespaceURI, measure.localName) not in validUtrUnits:
+                    customUnits.append(measure)
+    if customUnits:
+        for unit in customUnits:
+            yield Validation.warning(
+                codes='NL.FG-NL-11',
+                msg=_('Standard units SHOULD be used. "%(modelUnit)s" is not a standard unit.'),
+                modelObject=unit,
```
Comment (aaroncameron-wk) [inline arelle/plugin/validate/NL/rules/fg_nl.py:236]: Since you're adding measures (QNames) to `customUnits` above, `unit` here is actually a measure QName, not a ModelObject/ModelUnit. I'd suggest having `customUnits` being a list of tuples of `(unit: ModelUnit, measure: QName)` so you can then pass `unit` as `modelObject` and `measure` into the message itself.

File: arelle/plugin/validate/NL/rules/br_kvk.py [hunk 11]
```diff
@@ -134,17 +134,20 @@ def rule_br_kvk_3_01(
...
                 if measure.namespaceURI == 'http://www.xbrl.org/2003/iso4217':
                     currencyUnitIds.add(unitId)
                     currencyMeasures.append(measure)
+                    units.add(unit)
     if len(currencyMeasures) != 1:
         yield Validation.error(
             codes='NL.BR-KVK-3.01',
             msg=_('A measure element with a namespace prefix that refers to the "http://www.xbrl.org/2003/iso4217" '
                   'namespace MUST appear exactly once in the instance document. Units: %(unitIds)s, Measures: %(measures)s'),
+            modelObject=units,
```
Comment (derekgengenbacher-wf) [inline arelle/plugin/validate/NL/rules/br_kvk.py:150]: Why these changes?

Comment (allenbross-wf) [reply arelle/plugin/validate/NL/rules/br_kvk.py:150]: Aaron asked that I make them as we were looking at the code together.

Comment (aaroncameron-wk) [reply arelle/plugin/validate/NL/rules/br_kvk.py:150]: I noticed this validation didn't have refs previously.

**Review by aaroncameron-wk — APPROVED** [review]

File: arelle/plugin/validate/NL/rules/fg_nl.py [hunk 12]
```diff
@@ -193,3 +194,43 @@ def rule_fg_nl_09(
...
+    validUtrUnits = set()
+    for entries in utrItemTypeEntries.values():
+        for entry in entries.values():
+            if entry.nsUnit and entry.unitId:
+                validUtrUnits.add((entry.nsUnit, entry.unitId))
+    for unitId, unit in modelXbrl.units.items():
+        for measures in unit.measures:
+            for measure in measures:
+                if (measure.namespaceURI, measure.localName) not in validUtrUnits:
+                    customUnits.append((unit, measure))
```
Comment (austinmatherne-wk) [inline arelle/plugin/validate/NL/rules/fg_nl.py:228]: nit: instead of collecting custom units you could yield a validation warning here

**Review by austinmatherne-wk — APPROVED** [review]

**Review by austinmatherne-wk — APPROVED** [review]

**Review by derekgengenbacher-wf — APPROVED** [review]

**Comment by derekgengenbacher-wf** [comment]
QA +10 Testing locally succeeded.
